sw/device/tests/dif/dif_padctrl_test.cc

```diff
@@ -0,0 +1,756 @@
+      {PADCTRL_MIO_PADS6_REG_OFFSET, PADCTRL_MIO_PADS6_ATTR27_MASK,
+       PADCTRL_MIO_PADS6_ATTR27_OFFSET, 27},
+      {PADCTRL_MIO_PADS7_REG_OFFSET, PADCTRL_MIO_PADS7_ATTR28_MASK,
+       PADCTRL_MIO_PADS7_ATTR28_OFFSET, 28},
+      {PADCTRL_MIO_PADS7_REG_OFFSET, PADCTRL_MIO_PADS7_ATTR29_MASK,
+       PADCTRL_MIO_PADS7_ATTR29_OFFSET, 29},
+      {PADCTRL_MIO_PADS7_REG_OFFSET, PADCTRL_MIO_PADS7_ATTR30_MASK,
+       PADCTRL_MIO_PADS7_ATTR30_OFFSET, 30},
+      {PADCTRL_MIO_PADS7_REG_OFFSET, PADCTRL_MIO_PADS7_ATTR31_MASK,
+       PADCTRL_MIO_PADS7_ATTR31_OFFSET, 31}};
+
+  // Vector of all attributes, initialise with the basic attributes.
+  std::vector<dif_padctrl_attr_t> attributes_ = {
+      kDifPadctrlAttrIOInvert, kDifPadctrlAttrOpenDrain,
+      kDifPadctrlAttrPullDown, kDifPadctrlAttrPullUp,
+      kDifPadctrlAttrKeeper,   kDifPadctrlAttrWeakDrive};
+
+  /**
+   * Setup dummy `mmio_region_t` suitable for use with mock testing and
+   * initialise padctrl DIF using it.
+   */
+  void DeviceInit() {
+    mmio_region_t base_addr = dev().region();
+    EXPECT_EQ(dif_padctrl_init(base_addr, &dif_padctrl_), kDifPadctrlOk);
+  }
+
+  /**
+   * Expect a read of the REGEN register to check the lock bit
+   *
+   * @param locked Whether REGEN will report padctrl is locked or not
+   */
+  void ExpectLockCheck(bool locked) {
+    EXPECT_READ32(PADCTRL_REGEN_REG_OFFSET, locked ? 0 : 1);
+  }
+
+  /**
+   * Expect a sequence of writes that reset all attributes to the given
+   * `reset_val`
+   */
+  void ExpectReset(uint32_t reset_val) {
+    // Extract unique register offsets from dio_attr_fields_.
+    // set is required to keep ordering.
+    std::set<ptrdiff_t> dio_pad_reg_offsets;
+    for (auto dio_attr_field : dio_attr_fields_) {
+      dio_pad_reg_offsets.insert(dio_attr_field.reg_offset);
+    }
+
+    // Iterate through in ascending order, expecting a write of 0 to each.
+    for (auto dio_reg_offset : dio_pad_reg_offsets) {
+      EXPECT_WRITE32(dio_reg_offset, reset_val);
+    }
+
+    // Extract unique register offsets from mio_attr_fields_
+    std::set<ptrdiff_t> mio_pad_reg_offsets;
+    for (auto mio_attr_field : mio_attr_fields_) {
+      mio_pad_reg_offsets.insert(mio_attr_field.reg_offset);
+    }
+
+    // Iterate through in ascending order, expecting a write of 0 to each.
+    for (auto mio_reg_offset : mio_pad_reg_offsets) {
+      EXPECT_WRITE32(mio_reg_offset, reset_val);
+    }
+  }
+
+  /**
+   * Expects the read required to read an attribute field.
+   *
+   * @param attr_field Which attribute field we expect to read
+   * @param existing_attrs What existing attributes will be in the read
+   * @return The bits seen by the read
+   */
+  uint32_t ExpectAttrFieldRead(const AttrField &attr_field,
+                               dif_padctrl_attr_t existing_attrs) {
+    EXPECT_EQ(existing_attrs & ~attr_field.bit_mask, 0);
+
+    uint32_t other_field_bits = dev().GarbageMemory<uint32_t>() &
```

Comment:
Do not use `.GarbageMemory()` in user code. It is explicitly an implementation detail, and I *will* probably need to break this test in the future if you use it.

If `EXPECT_MASK` doesn't do what you want, please file an issue with the functionality you need added.

Reply:
I have left this as is. Padctrl is being updated to alter how pull-up and pull-down get set so that conflict cannot occur now. It is possible with this change we no longer require conflicting attribute checks in which case various things get simplified including this (https://github.com/lowRISC/opentitan/pull/2681). I'll wait to confirm we can remove conflicting attribute checks before looking at this.